### PR TITLE
Rewrite summarize configuration transparently

### PR DESCRIPTION
**Pull request by patszt**
Release 2.2 introduces changes to the summarize plugin config format.
VAST supports pre 2.2 config by internally formatting it to the new layout.

### :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [vast.io](https://vast.io), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

It was suggested to not unit test this code as it will be removed by the next release

### CODE REVIEW

**Comment by dominiklohmann** [comment]
> Mostly a question for @dominiklohmann: is there a place where we keep to-be-deprecated code that is scheduled to be removed in future major releases?

For both VAST itself and for the compaction plugin we have `try_handle_deprecations` functions; I suppose it'd make sense to use the same function name here as well. It should still be local to this plugin as it relates the the summarize plugin, not to VAST itself.

File: libvast/native-plugins/summarize.cpp [hunk 1]
```diff
@@ -750,10 +771,40 @@ class plugin final : public virtual pipeline_operator_plugin {
...
+                             fmt::format("Aggregate section found in legacy "
+                                         "config"));
+    }
+    // Reformat old config into new one.
+    auto new_config = config;
+    auto aggregate = record{};
+    for (const auto& section : sections_to_reformat) {
+      VAST_WARN("Legacy format detected in summarize section: {}. Please "
+                "upgrade to the newest format",
+                section);
```
Comment (dominiklohmann) [inline libvast/native-plugins/summarize.cpp]:
```suggestion
VAST_WARN("the configuration key '{}' for the summarize operator is deprecated and will be removed in the next major release; please migrate to using explicit output field names", section);
```

Modulo formatting.

File: libvast/native-plugins/summarize.cpp [hunk 2]
```diff
@@ -750,10 +796,22 @@ class plugin final : public virtual pipeline_operator_plugin {
...
+        sections_to_reformat.emplace_back(key);
+    }
+    // new format detected. Proceed as usual
+    if (sections_to_reformat.empty()) {
+      return make_summarize_operator(config);
+    }
+    if (config.contains("aggregate")) {
+      return caf::make_error(ec::invalid_configuration,
+                             fmt::format("Aggregate section found in legacy "
+                                         "config"));
```
Comment (dominiklohmann) [inline libvast/native-plugins/summarize.cpp:811]:
```suggestion
fmt::format("the configuration keys '{}' for the summarize operator are deprecated and will be removed in the next major release; please migrate to using explicit output field names", fmt::join(sections_to_reformat, "', '"));
```

File: libvast/native-plugins/summarize.cpp [hunk 3]
```diff
@@ -750,10 +796,22 @@ class plugin final : public virtual pipeline_operator_plugin {
...
-    return std::make_unique<summarize_operator>(std::move(*parsed_config));
+    std::vector<std::string> sections_to_reformat;
+    for (const auto& key : {"min", "max", "any", "all", "sum"}) {
+      if (config.contains(key))
+        sections_to_reformat.emplace_back(key);
+    }
+    // new format detected. Proceed as usual
+    if (sections_to_reformat.empty()) {
+      return make_summarize_operator(config);
+    }
```
Comment (dominiklohmann) [inline libvast/native-plugins/summarize.cpp:807]: Style nit: no braces for single-line scopes.

File: libvast/native-plugins/summarize.cpp [hunk 4]
```diff
@@ -735,6 +735,52 @@ class summarize_operator : public pipeline_operator {
...
+caf::expected<std::unique_ptr<pipeline_operator>>
+make_summarize_operator(const record& config) {
+  auto parsed_config = configuration::make(config);
+  if (!parsed_config)
+    return parsed_config.error();
+  return std::make_unique<summarize_operator>(std::move(*parsed_config));
+}
+
+std::vector<std::string>
+get_keys_from_legacy_sections(const data::variant& legacy_section) {
```
Comment (dominiklohmann) [inline libvast/native-plugins/summarize.cpp:747]:
```suggestion
get_keys_from_legacy_sections(const data& legacy_section) {
```

`data::variant` is an implementation detail that should not be needed here.

File: libvast/native-plugins/summarize.cpp [hunk 5]
```diff
@@ -735,6 +735,52 @@ class summarize_operator : public pipeline_operator {
...
+    VAST_WARN("Legacy format detected in summarize section: {}. Please "
+              "upgrade to the newest format",
+              section);
+    for (const auto& key :
+         get_keys_from_legacy_sections(new_config.at(section).get_data())) {
+      if (aggregate.contains(key))
+        return caf::make_error(ec::invalid_configuration,
+                               fmt::format("Duplicate key: {} found in "
+                                           "{} section",
+                                           key, section));
```
Comment (dominiklohmann) [inline libvast/native-plugins/summarize.cpp:775]:
```suggestion
                               fmt::format("aggregation function '{}' for output field '{}' shadows previous definition", section, key));
```

File: libvast/native-plugins/summarize.cpp [hunk 6]
```diff
@@ -735,6 +735,52 @@ class summarize_operator : public pipeline_operator {
...
+try_handle_deprecations(const record& config,
+                        const std::vector<std::string>& sections_to_reformat) {
+  auto new_config = config;
+  auto aggregate = record{};
+  for (const auto& section : sections_to_reformat) {
+    VAST_WARN("Legacy format detected in summarize section: {}. Please "
+              "upgrade to the newest format",
+              section);
+    for (const auto& key :
+         get_keys_from_legacy_sections(new_config.at(section).get_data())) {
```
Comment (dominiklohmann) [inline libvast/native-plugins/summarize.cpp:770]:
```suggestion
         get_keys_from_legacy_sections(new_config.at(section))) {
```

File: libvast/native-plugins/summarize.cpp [hunk 7]
```diff
@@ -750,10 +796,22 @@ class plugin final : public virtual pipeline_operator_plugin {
...
+    // new format detected. Proceed as usual
+    if (sections_to_reformat.empty()) {
+      return make_summarize_operator(config);
+    }
+    if (config.contains("aggregate")) {
+      return caf::make_error(ec::invalid_configuration,
+                             fmt::format("Aggregate section found in legacy "
+                                         "config"));
+    }
+
```
Comment (dominiklohmann) [inline libvast/native-plugins/summarize.cpp:813]: Style nit: unnecessary whitespace

**Review by dominiklohmann — APPROVED** [review]
This is great. I played around with this locally by reverting to an older version of our compactions used on our testbed, and they just worked with an appropriate warning.

File: libvast/native-plugins/summarize.cpp (hunk 1 above)
Comment (dominiklohmann) [reply libvast/native-plugins/summarize.cpp]: This was moved because there was a force-push in the middle of my review, but the comment still applies.

File: libvast/native-plugins/summarize.cpp (hunk 3 above)
Comment (patszt) [reply libvast/native-plugins/summarize.cpp:805]: I wonder if i will ever get used to this :D

File: libvast/native-plugins/summarize.cpp (hunk 1 above)
Comment (patszt) [reply libvast/native-plugins/summarize.cpp]: Done
